src/Simulation/qdk_sim_rs/src/c_api.rs

```diff
@@ -0,0 +1,443 @@
+    as_capi_err(|| apply(sim_id, &[idx], |model| &model.s))
+}
+
+/// Applies the `Adjoint S` operation acting on a given qubit to a given
+/// simulator, using the currently set noise model.
+#[no_mangle]
+pub fn s_adj(sim_id: usize, idx: usize) -> i64 {
+    as_capi_err(|| apply(sim_id, &[idx], |model| &model.s_adj))
+}
+
+/// Applies the `T` operation acting on a given qubit to a given simulator,
+/// using the currently set noise model.
+#[no_mangle]
+pub fn t(sim_id: usize, idx: usize) -> i64 {
+    as_capi_err(|| apply(sim_id, &[idx], |model| &model.t))
+}
+
+/// Applies the `Adjoint T` operation acting on a given qubit to a given simulator,
+/// using the currently set noise model.
+#[no_mangle]
+pub fn t_adj(sim_id: usize, idx: usize) -> i64 {
+    as_capi_err(|| apply(sim_id, &[idx], |model| &model.t_adj))
+}
+
+/// Applies the `CNOT` operation acting on two given qubits to a given simulator,
+/// using the currently set noise model.
+#[no_mangle]
+pub fn cnot(sim_id: usize, idx_control: usize, idx_target: usize) -> i64 {
+    as_capi_err(|| apply(sim_id, &[idx_control, idx_target], |model| &model.cnot))
+}
+
+/// Measures a single qubit in the $Z$-basis, returning the result by setting
+/// the value at a given pointer.
+///
+/// # Safety
+/// This function is marked as unsafe as it is the caller's responsibility to
+/// ensure that `result_out` is a valid pointer, and that the memory referenced
+/// by `result_out` can be safely set.
+#[no_mangle]
+pub unsafe extern "C" fn m(sim_id: usize, idx: usize, result_out: *mut usize) -> i64 {
+    as_capi_err(|| {
+        let state = &mut *STATE.lock().unwrap();
+        if let Some(sim_state) = state.get_mut(&sim_id) {
+            let instrument = &sim_state.noise_model.z_meas;
+            let (result, new_state) = instrument.sample(&[idx], &sim_state.register_state);
+            sim_state.register_state = new_state;
+            *result_out = result;
+            Ok(())
+        } else {
+            Err(format!("No simulator with id {} exists.", sim_id))
+        }
+    })
+}
+
+/// Gets the noise model corresponding to a particular name, serialized as a
+/// string representing a JSON object.
+///
+/// Currently recognized names:
+/// - `ideal`
+/// - `ideal_stabilizer`
+///
+/// # Safety
+/// As this is a C-language API, the Rust compiler cannot guarantee safety when
+/// calling into this function. The caller is responsible for ensuring that:
+///
+/// - `name` is a pointer to a null-terminated string of Unicode characters,
+///   encoded as UTF-8, and that the pointer remains valid for the lifetime of
+///   this call.
+/// - `noise_model_json` is a valid pointer whose lifetime extends for the
+///    duration of this function call.
+///
+/// After this call completes, this function guarantees that either of the two
+/// conditions below holds:
+///
+/// - The return value is negative, in which case calling `lasterr` will return
+///   an actionable error message, or
```

Comment:
(FYI for the future) In the `c-api.md` there is a fragment:
> Most C API functions for this crate return an integer, with `0` indicating success and **any other value indicating failure**.

The examples in `c-api.md` treat any non-zero return value (even _positive one_) as a failure.

If at some point in the future one wants to distinguish one group of success from the other group of success (e.g. "success with warning") by returning `0` vs. `1` (positive), then such a change can become a breaking change for a huge error-handling codebase.

You may want to consider one value for success (e.g. `0`) and one value for failure (e.g. `-1`), making _all other values reserved for the future_.

Reply:
https://github.com/microsoft/qsharp-runtime/issues/732